addons/purchase/tests/test_purchase_invoice.py

```diff
@@ -560,3 +590,80 @@ def test_on_change_quantity_price_unit(self):
         # Ensure price unit is updated when changing quantity on PO confirmed and line NOT linked to an invoice line
         po_line.write({'product_qty': 4})
         self.assertEqual(40.0, po_line.price_unit, "Unit price should be set to 40.0 for 4 quantity")
+
+
+@tagged('post_install', '-at_install')
+class TestInvoicePurchaseMatch(TestPurchaseToInvoiceCommon):
+
+    def test_total_match_via_partner(self):
+        po = self.init_purchase(confirm=True, partner=self.partner_a, products=[self.product_order])
+        invoice = self.init_invoice('in_invoice', partner=self.partner_a, products=[self.product_order])
+
+        invoice._find_and_set_purchase_orders(
+            [], invoice.partner_id.id, invoice.amount_total)
+
+        self.assertTrue(invoice.id in po.invoice_ids.ids)
+        self.assertEqual(invoice.amount_total, po.amount_total)
+
+    def test_total_match_via_po_reference(self):
+        po = self.init_purchase(confirm=True, products=[self.product_order])
+        invoice = self.init_invoice('in_invoice', partner=self.partner_a, products=[self.product_order])
+
+        invoice._find_and_set_purchase_orders(
+            ['my_match_reference'], invoice.partner_id.id, invoice.amount_total)
+
+        self.assertTrue(invoice.id in po.invoice_ids.ids)
+        self.assertEqual(invoice.amount_total, po.amount_total)
+
+    def test_subset_total_match_prefer_purchase(self):
+        po = self.init_purchase(confirm=True, products=[self.product_order, self.service_order])
+        invoice = self.init_invoice('in_invoice', partner=self.partner_a, products=[self.product_order])
+
+        invoice._find_and_set_purchase_orders(
+            ['my_match_reference'], invoice.partner_id.id, invoice.amount_total, prefer_purchase_line=True)
+        additional_unmatch_po_line = po.order_line.filtered(lambda _line: _line.product_id == self.service_order)
+
+        self.assertTrue(invoice.id in po.invoice_ids.ids)
+        self.assertEqual(len(invoice.line_ids), 6)
+        self.assertTrue(invoice.line_ids.filtered(
+            lambda _line: _line.purchase_line_id == additional_unmatch_po_line).quantity == 0)
```

Comment:
```suggestion
            lambda l: l.purchase_line_id == additional_unmatch_po_line).quantity == 0)
```
we can use single chars to distinguish lambda stuff